### DISCUSSION TITLE
Connector implementations in other programming languages | Oxidizing EDC

### DISCUSSION THREAD

**Discussion by St4NNi** [eclipse-edc/Connector#4313] [General]
Hello everyone,

First of all, thank you for your great work, we are very excited that the Dataspaces community has created so many useful components and tools. Currently, my team and I are working to integrate EDC and IDSA connector technologies with academic data providers to bridge the gap between industry and academic data sources.

I know this may be a controversial topic, but is there any interest in developing an EDC compatible connector in a programming language other than Java?  

We have been looking into developing a very lightweight compatible connector in Rust for quite some time now. And I know that developing multiple connectors in parallel is a huge undertaking, but there are a number of reasons why Java might not be a good fit for us:

- Java is dying in our academic context. It has been replaced by either Python and/or low-level system languages like C or Rust, so as a university we do not have a lot of human resources for it. And to be honest, it looks like this will only get worse in the future.
- The JVM (and even GraalVM) needs a lot of resources, which makes it quite difficult to run a connector in an embedded or edge use case.
- We would like to integrate connector technologies directly into other data management systems. But it seems there is no good way to embed (and greatly extend) a connector into an existing technology stack that interoperates with a C ABI without using the full connector (yes, there is JNI, but that is not really a good fit).
- The inherent memory safety, great performance, and predictability of a compiled Rust program seems like a perfect fit for a core infrastructure component like a dataspace connector.

There are other reasons against Java, but these should suffice to make our point. We have been developing bits and pieces for quite some time now (a working ODRL state machine, an EDC client, etc.), so we would not be starting from scratch, but there is a lot more to do. My main question / discussion would be: is there interest in this community? 

I would be very happy if there are a few other people interested in experimenting in this direction.

Thank you very much!

**Comment by paullatzelsperger** [comment]
Hi @St4NNi. 
In the enterprise context, which is the driving force behind EDC, and which is what EDC was developed for, Java is _very much_ a prevalent technology, because it's widely understood and known, it's very mature and stable and there is a huge community and great availability for it. So there is a myriad of reasons _for_ a JVM language as well, I just listed a few.

For those reasons we will be focussing solely on Java development for the forseeable future. 

However, EDC is based on open standards (DSP, DCP, ...) so if you like you can develop a Rust-based connector, and we'll be happy to list you as a "Friend of EDC" if your implementation fits our [adoption guidelines](https://github.com/eclipse-edc/docs/blob/main/contributing/adoption.md). 

Just so you know, there will be a [Compliance Verification Framework for DSP](https://github.com/eclipse-dataspacetck/cvf)(WIP at the moment) with which you can assert DSP compliance.

**Comment by ndr-brt** [comment]
A Rust implementation for the dataspace connector will be definitely something I'd like to sneak in, is there something already open sourced to look to?

**Reply by St4NNi** [reply]
We are currently collecting all the bits and pieces we have lying around into a more sophisticated repo, and plan to release our unfinished collection of libraries as open source very soon.

Please note that this is far from a fully functional connector, but it could be used as a starting point for building a minimal viable connector in rust.

**Reply by hansen-maria** [reply]
First of all, thank you for your interest and sorry for the late response.

We have now compiled all sub-projects in a Rust workspace. The public repository can be found at the following URL: [https://github.com/ArunaStorage/dspace-rs](https://github.com/ArunaStorage/dspace-rs)

But as Sebastian already mentioned, this is more of an initial start than a fully functional connector.
